### PR TITLE
version 2

### PR DESCRIPTION
**Pull request by ndem0**
- Move the python-occ to 'cad' submodule: from now on the 'python-occ' package is not mandatory for the installation, but only to access to this submodule. Tests are run according to the environment;
- Refactor the code, merging the parameter classes and the deformation ones (e.g. FFDParameters is encapsulated in FFD);
- Add `CustomDeformation` (both for CAD or discrete geometries);
- The Handlers classes for CAD are removed;
- The other Handlers are marked as deprecated (will be moved in next releases);
- Improvements in the CAD part, adding features to overcome the known issues.

### CODE REVIEW

File: pygem/cad/ffd.py [hunk 1]
```diff
@@ -173,76 +191,67 @@ def __call__(self, obj):
...
-            #IGESControl_Controller_Init()
-            #writer = IGESControl_Writer()
-            #writer.AddShape(brep)
-            #nomefile = "untrimmed_face"+str(faceCount)+".iges"
-            #writer.Write(nomefile)
+
             
             # we here start looping on the wires of the original face
-            # in this forst loop we do nothing but count the wires in this face
+            # in this first loop we do nothing but count the wires in this face
```
Comment (mtezzele) [inline pygem/cad/ffd.py:218]: I suppose this section of the code just prints informations. I would remove it.

File: pygem/cad/ffd.py [hunk 2]
```diff
@@ -183,52 +183,59 @@ def __call__(self, obj, dst=None):
...
             while wire_explorer.More():
                 wire = topods_Wire(wire_explorer.Current())
-                if (wire == breptools_OuterWire(face_aux)):
+                if wire == breptools_OuterWire(face_aux):
                     print("Wire", wire_count+1, "is outer wire")
                 wire_count += 1
                 wire_explorer.Next()
-            print("This face has ",wire_count," wires")
-            
+            print("This face has ", wire_count, " wires")
```
Comment (ndem0) [inline pygem/cad/ffd.py:233]: Remove the print

File: pygem/cad/ffd.py [hunk 3]
```diff
@@ -238,107 +245,117 @@ def __call__(self, obj, dst=None):
...
-                
+
                 bspline_geom_hedge = h_bspline_edge.BSplineCurve()
                 bspline_geom_edge = bspline_geom_hedge
-                unified_edge = BRepBuilderAPI_MakeEdge(bspline_geom_hedge).Edge()
 
-                # number of knots is enriched here: this can become a user prescribed parameter for the class
+                # number of knots is enriched here: this can become a user
+                # prescribed parameter for the class
                 knotsToAdd = 30
```
Comment (ndem0) [inline pygem/cad/ffd.py:272]: variable

File: pygem/cad/ffd.py [hunk 4]
```diff
@@ -238,107 +245,117 @@ def __call__(self, obj, dst=None):
...
                 firstParam = bspline_geom_edge.FirstParameter()
                 lastParam = bspline_geom_edge.LastParameter()
                 for i in range(knotsToAdd):
-                    bspline_geom_edge.InsertKnot(firstParam+i*(lastParam-firstParam)/knotsToAdd,1,1e-7)
+                    bspline_geom_edge.InsertKnot(firstParam+ \
+                               i*(lastParam-firstParam)/knotsToAdd, 1, 1e-7)
```
Comment (ndem0) [inline pygem/cad/ffd.py:277]: tolerance

File: pygem/cad/ffd.py [hunk 5]
```diff
@@ -238,107 +245,117 @@ def __call__(self, obj, dst=None):
...
                 shapesList = TopTools_ListOfShape()
                 # openCascade object
                 occ_edge = bspline_geom_edge
 
                 # extract the Control Points of each face
                 n_poles = occ_edge.NbPoles()
                 control_polygon_coordinates = np.zeros(\
-                shape=(n_poles , 3))
```
Comment (ndem0) [inline pygem/cad/ffd.py]: indentation

File: pygem/cad/ffd.py [hunk 6]
```diff
@@ -238,107 +245,117 @@ def __call__(self, obj, dst=None):
...
                 wire_maker.Add(shapesList)
                 result_wire = wire_maker.Wire()
                 iges_handler = IgesHandler()
-                iges_handler.write_shape_to_file(result_wire, "face_"+str(faceCount)+"_wire_"+str(wire_count)+".iges")
-
-                # now, the wire can be outer or inner. we store the outer and (possible) inner ones in different lists
-                # this is because we first need to trim the surface using the outer wire, and then we can trim it
-                # with the wires corresponding to all the holes. if this is not done, the procedure will not work
-                if (wire == breptools_OuterWire(face_aux)):
+                iges_handler.write_shape_to_file(result_wire, "face_"+ \
```
Comment (ndem0) [inline pygem/cad/ffd.py:318]: What are we saving to file actually?

**Comment by ndem0** [comment]
@andreamola Please, uncomment all the tests in file `tests/test_ffd.py`. I suggest to put the additional tests in a new test file.

File: pygem/cad/ffd.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+        >>> ffd = FFD()
+        >>> ffd.read_parameters(
+        >>>        'tests/test_datasets/parameters_test_ffd_iges.prm')
+        >>> input_cad_file_name = "input.iges"
+        >>> modified_cad_file_name = "output.iges"
+        >>> ffd(input_cad_file_name,modified_cad_file_name)
+    """
+    def __init__(self, n_control_points=None):
+        super().__init__(n_control_points=None)
+        self.uKnotsToAdd = 30
```
Comment (ndem0) [inline pygem/cad/ffd.py:107]: Missing information for `uKnotsToAdd` in class documentation

File: pygem/cad/ffd.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+        self.vKnotsToAdd = 30
```
Comment (ndem0) [inline pygem/cad/ffd.py:108]: Missing information for `vKnotsToAdd` in class documentation

File: pygem/cad/ffd.py [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+        self.knotsToAdd = 30
```
Comment (ndem0) [inline pygem/cad/ffd.py:109]: Missing information for `knotsToAdd` in class documentation

File: pygem/cad/ffd.py [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+        self.tolerance = 1e-4
```
Comment (ndem0) [inline pygem/cad/ffd.py:110]: Missing information for `tolerance` in class documentation

File: pygem/cad/ffd.py [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+        while faces_explorer.More():
+            # performing some conversions to get the right
+            # format (BSplineSurface)
+            face = topods_Face(faces_explorer.Current())
+            nurbs_converter = BRepBuilderAPI_NurbsConvert(face)
+            nurbs_face = nurbs_converter.Shape()
+            face_aux = topods_Face(nurbs_face)
+            brep_face = BRep_Tool.Surface(topods_Face(nurbs_face))
+            bounds = 0.0
+            bounds = brep_face.Bounds()
```
Comment (ndem0) [inline pygem/cad/ffd.py:149]: Class private method (one underscore in front of the name e.g. `_foo()`) for performing the cast. Describe input and output variables.

File: pygem/cad/ffd.py [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+
+            bspline_face = geomconvert_SurfaceToBSplineSurface(brep_face)
+            # we will then add an amount of nodes that will grant
+            # us our prescribed resolution both along u and v
+            for i in range(self.uKnotsToAdd):
+                bspline_face.InsertUKnot(bounds[0]+ \
+                    i*(bounds[1]-bounds[0])/self.uKnotsToAdd, 1, self.tolerance)
```
Comment (ndem0) [inline pygem/cad/ffd.py:156]: Code repetead three times. Please encapsulate it into a private method.

File: pygem/cad/ffd.py [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+            i = 0
+            for pole_u_direction in range(n_poles_u):
+                for pole_v_direction in range(n_poles_v):
+                    control_point = gp_Pnt(new_pts[i, 0],
+                                           new_pts[i, 1],
+                                           new_pts[i, 2])
+                    occ_face.SetPole(pole_u_direction + 1,
+                                     pole_v_direction + 1,
+                                     control_point)
+                    i += 1
```
Comment (ndem0) [inline pygem/cad/ffd.py:194]: Class private method (one underscore in front of the name e.g. `_foo()`). Describe input and output variables.

File: pygem/cad/ffd.py [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+            # is the outer, and the others are the inner ones.
+            # the loop will also tell us which wire is the outer one
+            wire_count = 0
+            wire_explorer = TopExp_Explorer(face_aux, TopAbs_WIRE)
+            while wire_explorer.More():
+                wire = topods_Wire(wire_explorer.Current())
+                if wire == breptools_OuterWire(face_aux):
+                    print("Wire", wire_count+1, "is outer wire")
+                wire_count += 1
+                wire_explorer.Next()
```
Comment (ndem0) [inline pygem/cad/ffd.py:220]: Purpose?

File: pygem/cad/ffd.py [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+                        bspline_converter = BRepBuilderAPI_NurbsConvert(edge)
+                        bspline_converter.Perform(edge)
+                        bspline_tshape_edge = bspline_converter.Shape()
+                        h_geom_edge, a, b = \
+                            BRep_Tool_Curve(topods_Edge(bspline_tshape_edge))
+                        this_bspline_edge = \
+                            geomconvert_CurveToBSplineCurve(h_geom_edge)
+                        bspline_geom_edge = this_bspline_edge
+                        h_bspline_edge.Add(this_bspline_edge, self.tolerance)
+                    edgesCount += 1
```
Comment (ndem0) [inline pygem/cad/ffd.py:250]: Class private method (one underscore in front of the name e.g. `_foo()`) for performing the cast. Describe input and output variables.

File: pygem/cad/ffd.py [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+                                           new_pts[i, 2])
+                    occ_edge.SetPole(pole + 1, control_point)
+                    i += 1
+
+                modified_edge = BRepBuilderAPI_MakeEdge(occ_edge).Edge()
+                shapesList.Append(modified_edge)
+
+                wire_maker = BRepBuilderAPI_MakeWire()
+                wire_maker.Add(shapesList)
+                result_wire = wire_maker.Wire()
```
Comment (ndem0) [inline pygem/cad/ffd.py:303]: Class private method (one underscore in front of the name e.g. `_foo()`). Describe input and output variables.

File: pygem/cad/ffd.py [hunk 17]
```diff
@@ -0,0 +1,435 @@
...
+            poles_coordinates[pole_id, :] = [pole.X(), pole.Y(), pole.Z()]
+
+
+        # the new poles positions are computed through FFD
+        new_pts = super().__call__(poles_coordinates)
+
+        # the Bspline curve is now looped again to
+        # set the poles positions  to new_points
+        i = 0
+        for pole in range(n_poles):
```
Comment (mtezzele) [inline pygem/cad/ffd.py:249]: you can avoid `i=0` and `i+=1` by just using the pole variable: `i` and `pole` are the same.

**Pull request merged**